Reference for stage 3 set-methods (#28414)
mdn/content: files/en-us/web/javascript/reference/global_objects/set/index.md

    </tr>
  </tbody>
</table>

To make them more generalizable, these methods don't just accept `Set` objects, but anything that's [set-like](#set-like_objects).

### Set-like objects

All [set methods](#set_methods) require {{jsxref("Operators/this", "this")}} to be an actual `Set` instance, but their arguments just need to be set-like. A _set-like object_ is an object that provides the following:

- A {{jsxref("Set/size", "size")}} property that contains a number.
- A {{jsxref("Set/has", "has()")}} method that takes an element and returns a boolean.
- A {{jsxref("Set/keys", "keys()")}} method that returns an [iterator](/en-US/docs/Web/JavaScript/Reference/Iteration_protocols) of the elements in the set.

For example, {{jsxref("Map")}} objects are set-like because they also have {{jsxref("Map/size", "size")}}, {{jsxref("Map/has", "has()")}}, and {{jsxref("Map/keys", "keys()")}}, so they behave just like sets of keys when used in set methods:

```js
const a = new Set([1, 2, 3]);
const b = new Map([
  [1, "one"],
  [2, "two"],
  [4, "four"],
]);
console.log(a.union(b)); // Set(4) {1, 2, 3, 4}
```

> > **Note:** The set-like protocol invokes the `keys()` method instead of [`[@@iterator]()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/@@iterator) to produce elements, because for maps, the iterator produces _entries_ but the `has()` method takes _keys_, and maps are intended to behave like sets of keys.

Reviewer: ```suggestion
> **Note:** The set-like protocol invokes the `keys()` method instead of [`[@@iterator]()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/@@iterator) to produce elements. For maps, the iterator produces _entries_ but the `has()` method takes _keys_ and maps are intended to behave like sets of keys.
```

  - : Removes all elements from the `Set` object.
- {{jsxref("Set.prototype.delete()")}}
  - : Removes the element associated to the `value` and returns a boolean asserting whether an element was successfully removed or not. `Set.prototype.has(value)` will return `false` afterwards.
- {{jsxref("Set.prototype.difference()")}}

>   - : Takes a set and returns a new set containing elements in this set but not in the given set.

Reviewer: ```suggestion
  - : Takes a set and returns a new set containing elements in this set that are not in the given set.
```

Author: I find both equally fine, but the preposition "but" makes the logical relationship slightly more saliant.

- {{jsxref("Set.prototype.entries()")}}
  - : Returns a new iterator object that contains **an array of `[value, value]`** for each element in the `Set` object, in insertion order. This is similar to the {{jsxref("Map")}} object, so that each entry's _key_ is the same as its _value_ for a `Set`.
- {{jsxref("Set.prototype.forEach()")}}
  - : Calls `callbackFn` once for each value present in the `Set` object, in insertion order. If a `thisArg` parameter is provided, it will be used as the `this` value for each invocation of `callbackFn`.
- {{jsxref("Set.prototype.has()")}}
  - : Returns a boolean asserting whether an element is present with the given value in the `Set` object or not.
- {{jsxref("Set.prototype.intersection()")}}

>   - : Takes a set and returns a new set containing elements both in this set and in the given set.

Reviewer: ```suggestion
  - : Takes a set and returns a new set containing elements in both this set and the given set.
```
